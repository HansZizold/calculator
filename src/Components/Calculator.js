import React, { useState } from 'react';
import styles from './Calculator.module.css';
import calculate from '../logic/calculate';

const Calculator = () => {
  const [data, setData] = useState({
    previousData: {
      total: null,
      next: null,
      operation: null,
    },
  });
  const [display, setDisplay] = useState('0');

  const handler = (event) => {
    const buttonValue = event.target.textContent;
    const actualData = calculate(data.previousData, buttonValue);
    const { total, next, operation } = actualData;
    let temp = '';
    if (total) {
      temp += total;
    }
    if (operation) {
      temp += operation;
    }
    if (next) {
      temp += next;
    }
    setData({ previousData: actualData });
    setDisplay(temp);
  };

  return (
    <div className={styles.maincalc}>
      <h2>Let&#39;s do some math!</h2>
      <div>
        <input type="text" value={display} className={styles.input} readOnly />
        <div className={styles.buttons1}>
          <button type="button" className={styles.button1} onClick={handler}>AC</button>
          <button type="button" className={styles.button1} onClick={handler}>+/-</button>
          <button type="button" className={styles.button1} onClick={handler}>%</button>
          <button type="button" className={styles.button2} onClick={handler}>รท</button>
        </div>
        <div className={styles.buttons1}>
          <button type="button" className={styles.button1} onClick={handler}>7</button>
          <button type="button" className={styles.button1} onClick={handler}>8</button>
          <button type="button" className={styles.button1} onClick={handler}>9</button>
          <button type="button" className={styles.button2} onClick={handler}>x</button>
        </div>
        <div className={styles.buttons1}>
          <button type="button" className={styles.button1} onClick={handler}>4</button>
          <button type="button" className={styles.button1} onClick={handler}>5</button>
          <button type="button" className={styles.button1} onClick={handler}>6</button>
          <button type="button" className={styles.button2} onClick={handler}>-</button>
        </div>
        <div className={styles.buttons1}>
          <button type="button" className={styles.button1} onClick={handler}>1</button>
          <button type="button" className={styles.button1} onClick={handler}>2</button>
          <button type="button" className={styles.button1} onClick={handler}>3</button>
          <button type="button" className={styles.button2} onClick={handler}>+</button>
        </div>
        <div className={styles.buttons1}>
          <button type="button" className={styles.button0} onClick={handler}>0</button>
          <button type="button" className={styles.button1} onClick={handler}>.</button>
          <button type="button" className={styles.button2} onClick={handler}>=</button>
        </div>
      </div>
    </div>
  );
};
export default Calculator;
