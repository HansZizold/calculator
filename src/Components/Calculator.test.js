import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import Calculator from './Calculator';

describe('Calculator User Interacion', () => {
  test('Adding', async () => {
    const user = userEvent.setup();
    render(<Calculator />);
    const inputElement = screen.getByRole('textbox');
    expect(inputElement.value).toBe('0');

    const number7 = screen.getByRole('button', { name: '7' });
    await user.click(number7);
    expect(inputElement.value).toBe('7');

    const sum = screen.getByRole('button', { name: '+' });
    await user.click(sum);
    expect(inputElement.value).toBe('7+');

    const number3 = screen.getByRole('button', { name: '3' });
    await user.click(number3);
    expect(inputElement.value).toBe('7+3');

    const equalBtn = screen.getByRole('button', { name: '=' });
    await user.click(equalBtn);
    expect(inputElement.value).toBe('10');
  });

  test('Dividing', async () => {
    const user = userEvent.setup();
    render(<Calculator />);
    const inputElement = screen.getByRole('textbox');
    expect(inputElement.value).toBe('0');

    let number7 = screen.getByRole('button', { name: '7' });
    await user.click(number7);
    expect(inputElement.value).toBe('7');

    number7 = screen.getByRole('button', { name: '7' });
    await user.click(number7);
    expect(inputElement.value).toBe('77');

    const div = screen.getByRole('button', { name: 'รท' });
    await user.click(div);
    expect(inputElement.value).toBe('77รท');

    number7 = screen.getByRole('button', { name: '7' });
    await user.click(number7);
    expect(inputElement.value).toBe('77รท7');

    const equal = screen.getByRole('button', { name: '=' });
    await user.click(equal);
    expect(inputElement.value).toBe('11');
  });
});

test('Calculator renders correctly', () => {
  const tree = renderer.create(<Calculator />).toJSON();
  expect(tree).toMatchSnapshot();
});
