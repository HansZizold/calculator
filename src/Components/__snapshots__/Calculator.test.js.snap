// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator renders correctly 1`] = `
<div
  className="maincalc"
>
  <h2>
    Let's do some math!
  </h2>
  <div>
    <input
      className="input"
      readOnly={true}
      type="text"
      value="0"
    />
    <div
      className="buttons1"
    >
      <button
        className="button1"
        onClick={[Function]}
        type="button"
      >
        AC
      </button>
      <button
        className="button1"
        onClick={[Function]}
        type="button"
      >
        +/-
      </button>
      <button
        className="button1"
        onClick={[Function]}
        type="button"
      >
        %
      </button>
      <button
        className="button2"
        onClick={[Function]}
        type="button"
      >
        รท
      </button>
    </div>
    <div
      className="buttons1"
    >
      <button
        className="button1"
        onClick={[Function]}
        type="button"
      >
        7
      </button>
      <button
        className="button1"
        onClick={[Function]}
        type="button"
      >
        8
      </button>
      <button
        className="button1"
        onClick={[Function]}
        type="button"
      >
        9
      </button>
      <button
        className="button2"
        onClick={[Function]}
        type="button"
      >
        x
      </button>
    </div>
    <div
      className="buttons1"
    >
      <button
        className="button1"
        onClick={[Function]}
        type="button"
      >
        4
      </button>
      <button
        className="button1"
        onClick={[Function]}
        type="button"
      >
        5
      </button>
      <button
        className="button1"
        onClick={[Function]}
        type="button"
      >
        6
      </button>
      <button
        className="button2"
        onClick={[Function]}
        type="button"
      >
        -
      </button>
    </div>
    <div
      className="buttons1"
    >
      <button
        className="button1"
        onClick={[Function]}
        type="button"
      >
        1
      </button>
      <button
        className="button1"
        onClick={[Function]}
        type="button"
      >
        2
      </button>
      <button
        className="button1"
        onClick={[Function]}
        type="button"
      >
        3
      </button>
      <button
        className="button2"
        onClick={[Function]}
        type="button"
      >
        +
      </button>
    </div>
    <div
      className="buttons1"
    >
      <button
        className="button0"
        onClick={[Function]}
        type="button"
      >
        0
      </button>
      <button
        className="button1"
        onClick={[Function]}
        type="button"
      >
        .
      </button>
      <button
        className="button2"
        onClick={[Function]}
        type="button"
      >
        =
      </button>
    </div>
  </div>
</div>
`;
